### PR TITLE
Paddle logcumsumexp

### LINKED ISSUE #19067: logcumsumexp

### PR DESCRIPTION
**Pull request by harrysharma1**
Close #19067

### CODE REVIEW

File: ivy/functional/ivy/experimental/elementwise.py [hunk 1]
```diff
@@ -1217,3 +1217,59 @@ def modf(
...
+    >>> ivy.logcumsumexp(data, axis=0)
+    [[ 0.        1.        2.        3.     ]
+    [ 4.01815   5.01815   6.01815   7.01815 ]
+    [ 8.018479  9.018479 10.018479 11.018479]]
+    >>> ivy.logcumsumexp(data, axis=1)
+    # [[ 0.         1.3132617  2.4076061  3.4401898]
+    #  [ 4.         5.3132615  6.407606   7.44019  ]
+    #  [ 8.         9.313262  10.407606  11.440189 ]]
+    """
+    return ivy.current_backend(x).logcumsumexp(x, axis=axis, dtype=dtype, out=out)
```
Comment (rishabgit) [inline ivy/functional/ivy/experimental/elementwise.py:1275]: Supported frameworks don't have `logcumsumexp` fn so this wouldn't work. E.g. if you search `def logcumsumexp`, you'd find it doesn't exist in any of the backends

**Review by rishabgit — CHANGES_REQUESTED** [review]
Hi @harrysharma1! Thanks for looking into this :smile: 
`logcumsumexp` doesn't exist in the backends, so current implementation wouldn't work.
I've added a comment in the exact place this is relevant, so it's easier for both of us to navigate. 
Hope these links are helpful - 
Relevant docs - https://unify.ai/docs/ivy/overview/deep_dive/ivy_frontends.html#ivy-frontends, https://unify.ai/docs/ivy/overview/deep_dive/function_types.html
Similar work - https://github.com/unifyai/ivy/blob/b09ad3db89d3f738463fdffe6d6ab5a50499ec9d/ivy/functional/frontends/torch/miscellaneous_ops.py#L187

Feel free to re-request review after making changes, thanks! 😄

**Review by rishabgit — CHANGES_REQUESTED** [review]
Hi, thanks a ton for making the changes! Could you please move the codes to paddle's frontend instead of adding it as a new fn in ivy's functional API? 

I think your current implementation is also on the right track, since pursing superset behaviour could make sense here. But we can always make that change later and for the scope of this task, I think it'd be cleaner to just have everything in the paddle frontend like the similar work for torch I linked earlier :smile:

**Comment by rishabgit** [comment]
Hey! There was a recent addition of backend_to_test arg in the testing pipeline, which was causing the code to fail tests. I added that, but it's throwing `AttributeError: module 'paddle.fluid.dygraph' has no attribute 'ParamBase'` now. It doesn't look to be due to your code, so I've dropped a note on our [discord](https://discord.com/channels/799879767196958751/1028267758028337193/1133662065974837289) regarding this. :smile: 
Until it's fixed -
- Could you take care of the merge conflicts ?
- It's best to avoid long variable names if shorter, intuitive alternates are available. Could you update that too? Might be good to just copy and paste the var names from the torch's frontend implementation, so it's easier to understand in the future.

Your code looks great otherwise btw! Thanks a lot of adding this :smile: 
If the above testing issue takes too long for us to fix, feel free to re-request another review and I'll merge your changes.

PS, I've been using pytest to check locally - 
`pytest /workspaces/ivy/ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_math.py::test_paddle_logcumsumexp`

**Review by rishabgit — CHANGES_REQUESTED** [review]
Hey! There was a recent addition of backend_to_test arg in the testing pipeline, which was causing the code to fail tests. I added that, but it's throwing `AttributeError: module 'paddle.fluid.dygraph' has no attribute 'ParamBase'` now. It doesn't look to be due to your code, so I've dropped a note on our [discord](https://discord.com/channels/799879767196958751/1028267758028337193/1133662065974837289) regarding this. :smile: 
Until it's fixed -
- Could you take care of the merge conflicts ?
- It's best to avoid long variable names if shorter, intuitive alternates are available. Could you update that too? Might be good to just copy and paste the var names from the torch's frontend implementation, so it's easier to understand in the future.

Your code looks great otherwise btw! Thanks a lot of adding this :smile: 
If the above testing issue takes too long for us to fix, feel free to re-request another review and I'll merge your changes.

PS, I've been using pytest to check locally - 
`pytest /workspaces/ivy/ivy_tests/test_ivy/test_frontends/test_paddle/test_tensor/test_math.py::test_paddle_logcumsumexp`

**Comment by rishabgit** [comment]
Update: looks like that testing issue was resolved by [this](https://github.com/unifyai/ivy/commit/0ea0f555fc3a989d2d2329c3a62110e7112c9b4c)
Could you please update your branch with latest changes?

**Comment by harrysharma1** [comment]
Screenshot 2023-07-26 at 11 57 28
This is causing paddle backend to act erratically so I cannot test it locally but i tried it previously and it worked and all i changed were the names to reflect the torch implementation.

File: ivy/functional/frontends/paddle/tensor/math.py [hunk 2]
```diff
@@ -352,6 +352,23 @@ def frac(x, name=None):
...
+    if len(x.shape) == 0:
+        ret = x
+    else:
+        # x is input
+        # name is out
+        original_dtype = dtype
+        exp_input = ivy.exp(x.astype("float64"))
+        summed_exp_input = ivy.cumsum(exp_input, axis=axis)
+        ret = ivy.log(summed_exp_input).astype(original_dtype)
+    if ivy.exists(name):
```
Comment (rishabgit) [inline ivy/functional/frontends/paddle/tensor/math.py:367]: not the right use of name - https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/logsumexp_en.html

File: ivy/functional/frontends/paddle/tensor/math.py [hunk 3]
```diff
@@ -352,6 +352,23 @@ def frac(x, name=None):
...
+        ivy.inplace_update(name, ret)
+    return ret
```
Comment (rishabgit) [inline ivy/functional/frontends/paddle/tensor/math.py:369]: doesn't return expected values - 
```
from ivy.functional.frontends.paddle import logcumsumexp
import ivy
ivy.set_backend("paddle")

x = ivy.array([[-1.5, 0., 2.], [3., 1.2, -2.4]])
print(logcumsumexp(x))

import paddle

x = paddle.to_tensor([[-1.5, 0., 2.], [3., 1.2, -2.4]])
print(paddle.logsumexp(x))
```

**Review by rishabgit — CHANGES_REQUESTED** [review]
Hey, feel free to re-request review after changes. Thanks! :smile:

File: ivy/functional/frontends/paddle/tensor/math.py (hunk 3 above)
Comment (harrysharma1) [reply ivy/functional/frontends/paddle/tensor/math.py:368]: I don't know if you realised but your showing an implementation of ```logsumexp``` on paddle but the ```logcumsumexp```  on the ivy version . I also added ```float16``` to supported data types. I tested it using the values on the paddle api docs and it returned with the correct values though with more decimal places :
Screenshot 2023-07-27 at 19 54 59
Screenshot 2023-07-27 at 19 55 20

**Comment by rishabgit** [comment]
Hi, I'm getting an odd `KeyError: 'cpu'` error. I don't think it's because of your code, something off with the testing pipeline. I've dropped a note to the testing team, will come back to your PR once it's resolved. Thanks! :smile: 
image

**Review by rishabgit — CHANGES_REQUESTED** [review]
Hi, the previous was fixed and no longer appears. But it's currently throwing `AssertionError: returned shape = (), ground-truth returned shape = (1,)`
Could you please take care of that? Don't forget to sync your branch with the latest updates! 

Feel free to re-request review afterward. Thanks! :smile:

**Comment by harrysharma1** [comment]
Hey I am still getting a key error on my local machine. I am not sure why

**Review by rishabgit — CHANGES_REQUESTED** [review]
Hey! That key error could be due to an old paddle installed on your local. Could you try updating paddle to 2.5.1?
After that, you should see the AssertionError I brought up earlier.

Feel free to re-request review after it's resolved. Thanks! :smile:
image

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**